[AST_generic] refactor type_parameter for Kotlin/Java/Scala/C++/Rust

aryx opened the pull request:
This will help PA-406
Lots of information was lost previously.

test plan:
make


PR checklist:
- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [ ] Change has security implications (if so, ping security team)

aryx committed:
[AST_generic] refactor type_parameter for Kotlin/Java/Scala/C++/Rust

This will help PA-406
Lots of information was lost previously.

test plan:
make

emjin commented:
Is it possible to add one or two test cases? At least, I would appreciate an example in the PR---I don't really understand what each field is for. (Trying to code review more!)

aryx committed:
Emma's comment

test plan:
You can look at the new tparams field below in the dumped AST.

```
yy -lang kotlin -dump_ast tests/kotlin/parsing/generics.kt
+ /home/pad/yy/_build/default/src/cli/Main.exe -lang kotlin -dump_ast tests/kotlin/parsing/generics.kt
[0.039  Info       Main.Dune__exe__Main ] loaded log_config.json
[0.039  Info       Main.Dune__exe__Main ] Executed as: /home/pad/yy/_build/default/src/cli/Main.exe -lang kotlin -dump_ast tests/kotlin/parsing/generics.kt
[0.039  Info       Main.Dune__exe__Main ] Version: semgrep-core version: v0.65.0-40-g1c11320b-dirty, pfff: 0.42
[0.039  Info       Main.Parse_target    ] trying to parse with TreeSitter parser tests/kotlin/parsing/generics.kt
Pr(
  [DefStmt(
     ({
       name=EN(
              Id(("Box", ()),
                {id_resolved=Ref(None); id_type=Ref(None);
                 id_constness=Ref(None); }));
       attrs=[];
       tparams=[{tp_id=("T", ()); tp_attrs=[]; tp_bounds=[]; tp_default=None;
                 tp_variance=None; tp_constraints=[]; }];
       },
      ClassDef(
        {ckind=(Class, ()); cextends=[]; cimplements=[]; cmixins=[];
         cparams=[ParamClassic(
                    {pname=Some(("t", ())); pdefault=None;
                     ptype=Some({t_attrs=[];
                                 t=TyTuple(
                                     [{t_attrs=[];
                                       t=TyN(
                                           Id(("T", ()),
                                             {id_resolved=Ref(None);
                                              id_type=Ref(None);
                                              id_constness=Ref(None); }));
                                       }]);
                                 });
                     pattrs=[];
                     pinfo={id_resolved=Ref(Some((Param, -1)));
                            id_type=Ref(None); id_constness=Ref(None); };
                     })];
         cbody=[FieldStmt(
                  DefStmt(
                    ({
                      name=EN(
                             Id(("value", ()),
                               {id_resolved=Ref(None); id_type=Ref(None);
                                id_constness=Ref(None); }));
                      attrs=[]; tparams=[]; },
                     VarDef(
                       {
                        vinit=Some(N(
                                     Id(("t", ()),
                                       {id_resolved=Ref(None);
                                        id_type=Ref(None);
                                        id_constness=Ref(None); })));
                        vtype=None; }))))];
         })));
   DefStmt(
     ({
       name=EN(
              Id(("Source", ()),
                {id_resolved=Ref(None); id_type=Ref(None);
                 id_constness=Ref(None); }));
       attrs=[];
       tparams=[{tp_id=("T", ()); tp_attrs=[]; tp_bounds=[]; tp_default=None;
                 tp_variance=Some((Covariant, ())); tp_constraints=[]; }];
       },
      ClassDef(
        {ckind=(Interface, ()); cextends=[]; cimplements=[]; cmixins=[
         ]; cparams=[];
         cbody=[FieldStmt(
                  DefStmt(
                    ({
                      name=EN(
                             Id(("nextT", ()),
                               {id_resolved=Ref(None); id_type=Ref(None);
                                id_constness=Ref(None); }));
                      attrs=[]; tparams=[]; },
                     FuncDef(
                       {fkind=(Function, ()); fparams=[];
                        frettype=Some({t_attrs=[];
                                       t=TyTuple(
                                           [{t_attrs=[];
                                             t=TyN(
                                                 Id(("T", ()),
                                                   {id_resolved=Ref(None);
                                                    id_type=Ref(None);
                                                    id_constness=Ref(
                                                    None); }));
                                             }]);
                                       });
                        fbody=FBDecl(()); }))))];
         })))])
```
